### PR TITLE
New bulk-transfer interface

### PR DESCRIPTION
**Pull request by benharsh**
This PR implements a new bulk transfer interface as described in GitHub Issue
#5310. The old interface was unnecessarily complex due to the various ways of
asking 'can we bulk transfer these arrays?'. This new interface attempts to be
simpler and easier to implement.

Users can now (optionally) implement any of the following methods:
```chpl
// srcClass and destClass inherit from BaseArr

// Transfers to/from well-understood distributions
proc srcClass.doiBulkTransferToKnown(srcDom:domain, destClass, destDom:domain) : bool {}
proc destClass.doiBulkTransferFromKnown(destDom:domain, srcClass, srcDom:domain) : bool {}

// Transfers to/from unknown distributions, likely through DefaultRectangulars
proc srcClass.doiBulkTransferToAny(srcDom:domain, destClass, destDom:domain) : bool {}
proc destClass.doiBulkTransferFromAny(destDom:domain, srcClass, srcDom:domain) : bool {}

// Opt-in to bulk-transfers involving rank changes
proc someArr.doiCanBulkTransferRankChange() param : bool {}
```

We check for bulk-transfer support by attempting to resolve these methods. If
we can resolve these methods, we will call the 'known' methods before the 'any'
methods. Methods return a bool indicating whether the bulk transfer occurred.
If the bulk-transfer did not occur, the caller (often internal modules) is
responsible for completing the transfer.

Testing:
- [x] full local
- [x] full gasnet, dist-cyclic, dist-replicated
- [x] numa, valgrind
- [x] chapcs playground performance
- [x] multilocale perf: ISx, miniMD, Stencil PRK

### CODE REVIEW

File: modules/internal/DefaultRectangular.chpl [hunk 1]
```diff
@@ -1690,70 +1697,95 @@ module DefaultRectangular {
...
     }
   }
 
-  proc DefaultRectangularArr.stridedTransferFrom(LViewDom, B) {
-    const SizeDims   = B.domain.dims();
-    const LHS        = this;
-    const RHS        = chpl__getActualArray(B);
-    const RViewDom   = chpl__getViewDom(B);
-    param targetRank = B.rank;
+  // Compute the active dimensions of this assignment. For example, LViewDom
```
Comment (daviditen) [inline modules/internal/DefaultRectangular.chpl]: This comment needs an update.  `LViewDom` and `RViewDom` have been renamed.

File: modules/dists/BlockDist.chpl [hunk 2]
```diff
@@ -1271,116 +1271,104 @@ proc BlockArr.dsiPrivatize(privatizeData) {
...
 
-proc BlockArr.doiCanBulkTransferStride(viewDom) param {
-  if debugBlockDistBulkTransfer then
-    writeln("In BlockArr.doiCanBulkTransferStride");
+  if A.sparseLayoutType != DefaultDist ||
+     A.sparseLayoutType != DefaultDist {
+    return false;
+  } else if aView.stridable || bView.stridable {
+    for param i in 1..A.rank {
+      if aView.dim(i).stride != 1 ||
```
[the remark this answers is not in the document]
Comment (benharsh) [reply modules/dists/BlockDist.chpl]: It is always true today, but I'll add a better error message anyways.

File: modules/dists/BlockDist.chpl [hunk 3]
```diff
@@ -1479,125 +1467,136 @@ proc dropDims(D: domain, dims...) {
...
+            r1[t] = (ini[t]:el..end[t]:el by sb[t]:el);
+            if r1[t].length != r2[t].length then
+              r1[t] = (ini[t]:el..end[t]:el by (end[t] - ini[t]):el/(r2[t].length-1));
         }
+
+        if debugBlockDistBulkTransfer then
+          writeln("A.locArr[i][", regionDest, "] = B[", (...r1), "]");
+
+        // TODO: handle possibility that this function returns false
+        chpl__bulkTransferArray(Dest.locArr[i].myElems._value, regionDest, Src, {(...r1)});
```
[the remark this answers is not in the document]
Comment (benharsh) [reply modules/dists/BlockDist.chpl]: I see this in vim as well. I think it's just that our syntax highlighters are basically ripped from C, and don't handle this (valid) case.

File: modules/dists/BlockDist.chpl [hunk 4]
```diff
@@ -1479,125 +1467,136 @@ proc dropDims(D: domain, dims...) {
...
+        const sb  = chpl__tuplify(regionSrc.stride);
+
+        var r1,r2: rank * range(idxType = el,stridable = true);
+        r2 = regionDest.dims();
+         //In the case that the number of elements in dimension t for r1 and r2
+         //were different, we need to calculate the correct stride in r1
+        for param t in 1..rank {
+            r1[t] = (ini[t]:el..end[t]:el by sb[t]:el);
+            if r1[t].length != r2[t].length then
+              r1[t] = (ini[t]:el..end[t]:el by (end[t] - ini[t]):el/(r2[t].length-1));
```
[the remark this answers is not in the document]
Comment (benharsh) [reply modules/dists/BlockDist.chpl]: This code was written a long time ago, I've just copied/moved it around for this PR. I'll try cleaning up sections like this.

File: modules/internal/ChapelArray.chpl [hunk 5]
```diff
@@ -3673,46 +3604,71 @@ module ChapelArray {
...
+  }
+  inline proc chpl__bulkTransferArray(ref a: [], AD : domain, const ref b: [], BD : domain) {
+    return chpl__bulkTransferArray(a._value, AD, b._value, BD);
+  }
+
+  inline proc chpl__bulkTransferArray(destClass, destDom : domain, srcClass, srcDom : domain) {
+    use Reflection;
+    var success = false;
+
+    proc bulkTransferDebug(msg:string) {
```
[the remark this answers is not in the document]
Comment (benharsh) [reply modules/internal/ChapelArray.chpl]: Good catch, I'll inline it.

File: modules/internal/ChapelArray.chpl [hunk 6]
```diff
@@ -3673,46 +3604,71 @@ module ChapelArray {
...
+    bulkTransferDebug("in chpl__bulkTransferArray");
+
+    //
+    // BHARSH TODO: I would prefer to hoist these 'canResolveMethod' calls into
+    // param bools before the if/else chain, but the compiler complains about
+    // hitting the instantiation limit for 'canResolveMethod'...
+    //
+    // TODO: should we attempt other bulk transfer methods if one fails?
+    //
+    if canResolveMethod(destClass, "doiBulkTransferFromKnown", destDom, srcClass, srcDom) {
```
[the remark this answers is not in the document]
Comment (benharsh) [reply modules/internal/ChapelArray.chpl]: I'm not sure, and I don't feel strongly one way or another. I left it as it is now to be safe.

**Comment by benharsh** [comment]
I've pushed some changes, and will start re-testing some configs.

**Comment by benharsh** [comment]
Testing came back clean.

**Pull request merged**